src/SMPProducerConsumer.java

```diff
@@ -0,0 +1,190 @@
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+import java.util.concurrent.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class SMPProducerConsumer {
+
+    private int[][] pref1;
+    private int[][] pref2;
+    private int[] proposed;
+    private boolean[][] assignment;
+
+
+    public SMPProducerConsumer(int[][] man_preferences, int[][] women_preferences, String optimality) {
+        if (optimality.compareTo("m") == 0) {
+            this.pref1 = man_preferences;
+            this.pref2 = women_preferences;
+        } else {
+            this.pref1 = women_preferences;
+            this.pref2 = man_preferences;
+        }
+
+        int count = man_preferences.length;
+
+        proposed = new int[count];
+        assignment = new boolean[count][count];
+    }
+
+    class Worker implements Runnable {
+
+        private ConcurrentLinkedQueue<Integer> _queue;
+
+        public Worker(ConcurrentLinkedQueue<Integer> queue) {
+            _queue = queue;
+        }
+
+        @Override
+        public void run() {
+            Integer freePerson;
+            while ((freePerson = _queue.poll()) != null) {
+                int preferredPerson = findPreferredPerson(freePerson);
+
+                synchronized (assignment) {
```

Comment:
This of course can be optimized since this is holding a lock for the ` assignment`  object during the whole execution of proposing, rejecting and finally enqueuing back the rejected person. This is an area of improvement.